### PR TITLE
Feature auth

### PR DESCRIPTION
**Pull request by Shundaa**

### CODE REVIEW

File: src/main/java/com/wipro/api/controller/IndexController.java [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+
+	@GetMapping(value = "/health-check/database", produces = "application/json")
+	public ResponseEntity<String> db() {
+		return new ResponseEntity<>(oracleConnection.getHealthDB(), HttpStatus.OK);
+	}
+
+	@PostMapping(value = "/auth", produces = "application/json", consumes = "application/json")
+	public ResponseEntity<String> auth(@RequestBody Message message) {
+		message = validationService.validation(message);
+		return new ResponseEntity<>(message.toString(), HttpStatus.OK);
```
Comment (victorgilc) [inline src/main/java/com/wipro/api/controller/IndexController.java]: Why toString here?

File: src/main/java/com/wipro/api/exception/ApiExceptionHandler.java [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+	private HttpHeaders headers;
+
+	@PostConstruct
+	public void init() {
+		headers = new HttpHeaders();
+	}
+
+	@ExceptionHandler
+	@ResponseBody
+	public ResponseEntity<String> handleDataBaseError(JDBCConnectionException ex) {
```
Comment (victorgilc) [inline src/main/java/com/wipro/api/exception/ApiExceptionHandler.java]: You can return a JsonNode

File: src/main/java/com/wipro/api/exception/ApiExceptionHandler.java [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+		return new ResponseEntity<String>("{\"Message\":\"Up and running Database Fail\"}", headers,
+				HttpStatus.SERVICE_UNAVAILABLE);
+	}
+
+	@ExceptionHandler
+	@ResponseBody
+	public ResponseEntity<String> handleValidationError(ValidationException validation) {
+		headers.setContentType(MediaType.APPLICATION_JSON);
+		// Remove, but will be on the nexts update
+		// StringBuilder error = new StringBuilder();
```
Comment (victorgilc) [inline src/main/java/com/wipro/api/exception/ApiExceptionHandler.java]: Avoid unnecessary comments

File: src/main/java/com/wipro/api/exception/ApiExceptionHandler.java [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+
+	@ExceptionHandler
+	@ResponseBody
+	public ResponseEntity<String> handleValidationError(ValidationException validation) {
+		headers.setContentType(MediaType.APPLICATION_JSON);
+		// Remove, but will be on the nexts update
+		// StringBuilder error = new StringBuilder();  
+		// validation.getList().forEach(errorsMessage ->
+		// error.append(",\""+errorsMessage.getErrorMessage()+"\""));
+		return new ResponseEntity<String>("{\"message\":\"Error\"" + ",\"erroCode:\":\"4002\""
```
Comment (victorgilc) [inline src/main/java/com/wipro/api/exception/ApiExceptionHandler.java]: JsonNode here, please

File: src/main/java/com/wipro/api/exception/ValidationException.java [hunk 5]
```diff
@@ -0,0 +1,25 @@
+package com.wipro.api.exception;
+
+import java.util.List;
+import com.wipro.api.model.ErrorMessageInvalid;
+
+public class ValidationException extends RuntimeException {
+	/**
+	 *
```
Comment (victorgilc) [inline src/main/java/com/wipro/api/exception/ValidationException.java:13]: Never use Serializable without a valuable UID.

File: src/main/java/com/wipro/api/exception/ValidationException.java [hunk 6]
```diff
@@ -0,0 +1,25 @@
...
+	 */
+	private static final long serialVersionUID = 1L;
+
+	private List<ErrorMessageInvalid> list;
+
+	public ValidationException(List<ErrorMessageInvalid> list) {
+		this.list = list;
+	}
+
+	public List<ErrorMessageInvalid> getList() {
```
Comment (victorgilc) [inline src/main/java/com/wipro/api/exception/ValidationException.java]: Use Lombok for it

File: src/main/java/com/wipro/api/model/ErrorMessageInvalid.java [hunk 7]
```diff
@@ -0,0 +1,15 @@
+package com.wipro.api.model;
+
+public class ErrorMessageInvalid {
+
+	private String errorMessage;
+
+	public ErrorMessageInvalid(String errorMessage) {
+		this.errorMessage = errorMessage + " Invalid";
```
Comment (victorgilc) [inline src/main/java/com/wipro/api/model/ErrorMessageInvalid.java]: StringBuilder to concatenation

File: src/main/java/com/wipro/api/model/ErrorMessageInvalid.java [hunk 8]
```diff
@@ -0,0 +1,15 @@
...
+	}
+
+	public String getErrorMessage() {
```
Comment (victorgilc) [inline src/main/java/com/wipro/api/model/ErrorMessageInvalid.java]: Lombok here

File: src/main/java/com/wipro/api/model/Message.java [hunk 9]
```diff
@@ -0,0 +1,49 @@
+package com.wipro.api.model;
+
+import com.google.gson.JsonObject;
+
+public class Message {
```
Comment (victorgilc) [inline src/main/java/com/wipro/api/model/Message.java:9]: Lombok everywhere

File: src/main/java/com/wipro/api/services/ConnectionService.java [hunk 10]
```diff
@@ -0,0 +1,26 @@
...
+public class ConnectionService {
+
+	@Autowired
+	private EntityManager entityManager;
+
+	@Value("${spring.queries.status-query}")
+	private String query;
+
+	public String getHealth() {
+		return "{\"message\":\"Up and running\"}";
```
Comment (victorgilc) [inline src/main/java/com/wipro/api/services/ConnectionService.java]: JsonNode here

File: src/main/java/com/wipro/api/services/ConnectionService.java [hunk 11]
```diff
@@ -0,0 +1,26 @@
...
+	}
+
+	public String getHealthDB() {
+		return entityManager.createNativeQuery(query).getSingleResult().toString();
```
Comment (victorgilc) [inline src/main/java/com/wipro/api/services/ConnectionService.java]: Why String here?

File: src/main/java/com/wipro/api/services/ValidationService.java [hunk 12]
```diff
@@ -0,0 +1,46 @@
...
+import org.springframework.stereotype.Service;
+
+import com.wipro.api.exception.ValidationException;
+import com.wipro.api.model.ErrorMessageInvalid;
+import com.wipro.api.model.Message;
+
+@Service
+public class ValidationService {
+	public Message validation(Message message) {
+		ArrayList<ErrorMessageInvalid> error = new ArrayList<>();
```
Comment (victorgilc) [inline src/main/java/com/wipro/api/services/ValidationService.java]: Avoid use ArrayList on the left side, use a List, for example, is all that you need.

File: src/main/java/com/wipro/api/services/ValidationService.java [hunk 13]
```diff
@@ -0,0 +1,46 @@
...
+		messageTypeValidation(message, error);
+		messageCategoryValidation(message, error);
+		deviceChannelValidation(message, error);
+		if (!error.isEmpty())
```
Comment (victorgilc) [inline src/main/java/com/wipro/api/services/ValidationService.java]: Use keys to increase readability

File: src/main/java/com/wipro/api/services/ValidationService.java [hunk 14]
```diff
@@ -0,0 +1,46 @@
...
+			throw new ValidationException(error);
+		message.setMessageType("ARes");
```
Comment (victorgilc) [inline src/main/java/com/wipro/api/services/ValidationService.java]: Set this as a constant

File: src/main/java/com/wipro/api/services/ValidationService.java [hunk 15]
```diff
@@ -0,0 +1,46 @@
...
+		return message;
+	}
+
+	private void messageTypeValidation(Message message, ArrayList<ErrorMessageInvalid> validation) {
```
Comment (victorgilc) [inline src/main/java/com/wipro/api/services/ValidationService.java]: You don't need ArrayList here

File: src/main/java/com/wipro/api/services/ValidationService.java [hunk 16]
```diff
@@ -0,0 +1,46 @@
...
+		if (!(message.getMessageType() == null
+				|| (Pattern.matches("[ARP]{1}[R]{1}[e]{1}[qs]{1}", message.getMessageType())
```
Comment (victorgilc) [inline src/main/java/com/wipro/api/services/ValidationService.java]: Break it, hard do read

File: src/main/java/com/wipro/api/services/ValidationService.java [hunk 17]
```diff
@@ -0,0 +1,46 @@
...
+						|| Pattern.matches("Erro", message.getMessageType())))) {
+			validation.add(new ErrorMessageInvalid("messageType"));
```
Comment (victorgilc) [inline src/main/java/com/wipro/api/services/ValidationService.java]: messageType can be a constant or enum

File: src/main/java/com/wipro/api/services/ValidationService.java [hunk 18]
```diff
@@ -0,0 +1,46 @@
...
+		}
+	}
+
+	private void messageCategoryValidation(Message message, ArrayList<ErrorMessageInvalid> validation) {
+		if (!(message.getMessageCategory() == null
```
Comment (victorgilc) [inline src/main/java/com/wipro/api/services/ValidationService.java]: Break it, hard do read

File: src/main/java/com/wipro/api/services/ValidationService.java [hunk 19]
```diff
@@ -0,0 +1,46 @@
...
+				|| Pattern.matches("[0]{1}[1-2]{1}", message.getMessageCategory()))) {
+			validation.add(new ErrorMessageInvalid("nessageCategory"));
+		}
+	}
+
+	private void deviceChannelValidation(Message message, ArrayList<ErrorMessageInvalid> validation) {
+		if (!(message.getDeviceChannel() == null || Pattern.matches("[0]{1}[1-3]{1}", message.getDeviceChannel()))) {
```
Comment (victorgilc) [inline src/main/java/com/wipro/api/services/ValidationService.java]: Break it, hard do read

File: src/main/java/com/wipro/api/services/ValidationService.java [hunk 20]
```diff
@@ -0,0 +1,46 @@
...
+	}
+
+	private void messageCategoryValidation(Message message, ArrayList<ErrorMessageInvalid> validation) {
+		if (!(message.getMessageCategory() == null
+				|| Pattern.matches("[0]{1}[1-2]{1}", message.getMessageCategory()))) {
+			validation.add(new ErrorMessageInvalid("nessageCategory"));
+		}
+	}
+
+	private void deviceChannelValidation(Message message, ArrayList<ErrorMessageInvalid> validation) {
```
Comment (victorgilc) [inline src/main/java/com/wipro/api/services/ValidationService.java]: ArrayList here again, replace please

File: src/test/java/com/wipro/api/controller/IndexControllerTest.java [hunk 21]
```diff
@@ -0,0 +1,59 @@
...
+import com.wipro.api.services.ConnectionService;
+import com.wipro.api.services.ValidationService;
+
+@RunWith(MockitoJUnitRunner.class)
+public class IndexControllerTest {
+
+	@InjectMocks
+	private IndexController indexControler;
+
+	@InjectMocks
```
Comment (victorgilc) [inline src/test/java/com/wipro/api/controller/IndexControllerTest.java]: Hardly you need to use more than one InjectMocks per Test, why this?

File: src/test/java/com/wipro/api/services/ConnectionServiceTest.java [hunk 22]
```diff
@@ -0,0 +1,37 @@
...
+import org.mockito.runners.MockitoJUnitRunner;
+
+@RunWith(MockitoJUnitRunner.class)
+public class ConnectionServiceTest {
+
+	@InjectMocks
+	ConnectionService connectionService;
+
+	@Mock
+	EntityManager entityManager;
```
Comment (victorgilc) [inline src/test/java/com/wipro/api/services/ConnectionServiceTest.java]: private here

File: src/test/java/com/wipro/api/services/ConnectionServiceTest.java [hunk 23]
```diff
@@ -0,0 +1,37 @@
...
+import org.mockito.InjectMocks;
+import org.mockito.Mock;
+import org.mockito.Mockito;
+import org.mockito.runners.MockitoJUnitRunner;
+
+@RunWith(MockitoJUnitRunner.class)
+public class ConnectionServiceTest {
+
+	@InjectMocks
+	ConnectionService connectionService;
```
Comment (victorgilc) [inline src/test/java/com/wipro/api/services/ConnectionServiceTest.java]: private here

File: src/test/java/com/wipro/api/services/ValidationServiceTest.java [hunk 24]
```diff
@@ -0,0 +1,96 @@
...
+import com.wipro.api.exception.ValidationException;
+import com.wipro.api.model.Message;
+
+@RunWith(MockitoJUnitRunner.class)
+public class ValidationServiceTest {
+
+	@Rule
+	public ExpectedException exception = ExpectedException.none();
+
+	@InjectMocks
```
Comment (victorgilc) [inline src/test/java/com/wipro/api/services/ValidationServiceTest.java:21]: Hardly you need to use more than one InjectMocks per Test, why this?

File: src/test/resources/queries.properties [hunk 25]
```diff
@@ -0,0 +1 @@
+spring.queries.status-query=SELECT '{\"message\":\"Up and running Database\"}' FROM dual
```
Comment (victorgilc) [inline src/test/resources/queries.properties]: What's going on here? Please remove this JSON and build this on java

File: src/test/java/com/wipro/api/services/ConnectionServiceTest.java [hunk 26]
```diff
@@ -0,0 +1,37 @@
...
+
+	@Mock
+	EntityManager entityManager;
+
+	@Test
+	public void connectionService_UpAndRunning() {
+		Assert.assertEquals(connectionService.getHealth(), "{\"message\":\"Up and running\"}");
```
Comment (victorgilc) [inline src/test/java/com/wipro/api/services/ConnectionServiceTest.java]: Crazy json, replace it

File: src/test/java/com/wipro/api/controller/IndexControllerTest.java [hunk 27]
```diff
@@ -0,0 +1,59 @@
...
+	private ConnectionService connectionService;
+
+	@Mock
+	private ValidationService validationService;
+
+	@Rule
+	public ExpectedException exception;
+
+	@Test
+	public void IndexController_HealthCheckReturnsOK() {
```
Comment (victorgilc) [inline src/test/java/com/wipro/api/controller/IndexControllerTest.java]: method name always starts with a lowercase letter

File: src/main/resources/queries.properties [hunk 28]
```diff
@@ -0,0 +1 @@
+spring.queries.status-query=SELECT '{\"message\":\"Up and running Database\"}' FROM dual
```
Comment (victorgilc) [inline src/main/resources/queries.properties]: Crazy json, replace it

File: src/main/java/com/wipro/api/model/Message.java [hunk 29]
```diff
@@ -0,0 +1,49 @@
...
+		return messageCategory;
+	}
+
+	public void setMessageCategory(String messageCategory) {
+		this.messageCategory = messageCategory;
+	}
+
+	@Override
+	public String toString() {
+		JsonObject message = new JsonObject();
```
Comment (victorgilc) [inline src/main/java/com/wipro/api/model/Message.java]: Use JsonNode. Can you use something else instead toString to convert it, please

**Review by victorgilc — COMMENTED** [review]
Reviewed. Please fix

**Review by GuzztMega — APPROVED** [review]

**Pull request merged**